File: app/src/main/java/com/xojiakbar/taskmanager/Utils/BaseViewModel.java
package com.xojiakbar.taskmanager.Utils;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.AndroidViewModel;

import java.lang.ref.WeakReference;

public abstract class BaseViewModel<R extends BaseEmptyRouter> extends AndroidViewModel {


    private WeakReference<R> router;
    Application application;



    public BaseViewModel(@NonNull Application application) {
        super(application);
        this.application = application;

        init();
    }

    protected void init() {

    }
    @Nullable
    public R getRouter() {
        if (router != null)
            return router.get();
        else
            return null;
    }

    public void setRouter(R r) {
        router = new WeakReference<>(r);
    }


}

File: app/src/main/java/com/xojiakbar/taskmanager/ui/base/BaseViewModel.kt
package com.xojiakbar.taskmanager.ui.base

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import java.lang.ref.WeakReference

abstract class BaseViewModel<R : BaseEmptyRouter?>(var application: Application) :
    AndroidViewModel(application) {
    private var router: WeakReference<R>? = null
//    var serialExecuter: SerialExecuter
//
//    init {
//        serialExecuter = SerialExecuter()
//        init()
//    }

    protected fun init() {}
    fun getRouter(): R? {
        return if (router != null) router!!.get() else null
    }

    fun setRouter(r: Nothing?) {
        router = WeakReference(r)
    }
}